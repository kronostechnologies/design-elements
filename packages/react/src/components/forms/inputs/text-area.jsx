import React, { Component } from 'react';
import styled from 'styled-components';

import style from '../styles/inputs';

import FieldContainer from '../field-container';

const StyledTextArea = styled.textarea`
  ${style}
  min-height: 6.5rem;
  min-width: 100%;
  outline: none;
  overflow: auto;
  resize: vertical;
`;

class TextArea extends Component {
    constructor(props) {
        super(props);

        this.state = {
            validity: true,
        };

        this.handleCheckValidity = this.handleCheckValidity.bind(this);
    }

    handleCheckValidity(thatEvt) {
        return this.setState({ validity: thatEvt.target.checkValidity() });
    }

    render() {
        const { id, label, required, valid, validMsg, ...props } = this.props;
        const { validity } = this.state;
        const isValid = (valid === undefined ? validity : valid);

        return (
            <FieldContainer
                fieldId={id}
                label={label}
                valid={isValid}
                validMsg={validMsg || 'This text area input is invalid'}
            >
                <StyledTextArea
                    {...props}
                    id={id}
                    onBlur={thatEvt => this.handleCheckValidity(thatEvt)}
                    required={required}
                />
            </FieldContainer>
        );
    }
}

export default TextArea;
