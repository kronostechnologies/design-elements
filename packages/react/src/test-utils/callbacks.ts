export function doNothing(): void {
    // do nothing
}
