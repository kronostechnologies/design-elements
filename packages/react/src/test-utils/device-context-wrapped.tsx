import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
