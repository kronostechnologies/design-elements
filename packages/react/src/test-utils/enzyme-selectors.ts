import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
