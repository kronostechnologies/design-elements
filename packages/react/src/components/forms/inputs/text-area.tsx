import React, { ChangeEvent, useState } from 'react';
import uuid from 'uuid/v4';

import styled from 'styled-components';
import { FieldContainer } from '../field-container';
import { inputsStyle } from '../styles/inputs';

const StyledTextArea = styled.textarea`
  ${inputsStyle}
  min-height: 6.5rem;
  min-width: 100%;
  outline: none;
  overflow: auto;
  resize: vertical;
`;

export interface TextAreaProps {
    defaultValue?: string;
    disabled?: boolean;
    label: string;
    required?: boolean;
    validMsg?: string;
    blurCallback?(value: string): void;
    changeCallback?(value: string): void;
    focusCallback?(value: string): void;
}

interface ValueProps { value: string; }
interface ValidityProps { validity: boolean; }

const TextArea = ({ defaultValue, disabled, label, blurCallback, changeCallback, focusCallback, required, validMsg }: TextAreaProps) => {
    const [{ value }, setValue] = useState<ValueProps>({ value: defaultValue || '' });
    const [{ validity }, setValidity] = useState<ValidityProps>({ validity: true });
    const id = uuid();

    const handleBlur = (event: any) => {
        const newValue = event.target.value;

        setValue({ value: newValue });
        setValidity({ validity: event.target.checkValidity() });

        if (blurCallback) {
            blurCallback(newValue);
        }
    };

    const handleChange = (event: ChangeEvent<HTMLTextAreaElement>) => {
        const newValue = event.target.value;
        setValue({ value: newValue });

        if (changeCallback) {
            changeCallback(newValue);
        }
    };

    const handleFocus = () => {
        if (focusCallback) {
            focusCallback(value);
        }
    };

    return (
        <FieldContainer
            fieldId={id}
            label={label}
            valid={validity}
            validMsg={validMsg || 'This text area input is invalid'}
        >
            <StyledTextArea
                disabled={disabled}
                id={id}
                onBlur={handleBlur}
                onChange={handleChange}
                onFocus={handleFocus}
                required={required}
                value={value}
            />
        </FieldContainer>
    );
};

export { TextArea };
