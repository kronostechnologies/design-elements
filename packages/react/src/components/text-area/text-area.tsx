import React, { ChangeEvent, FocusEvent, ReactElement, useState } from 'react';

import styled from 'styled-components';
import uuid from 'uuid/v4';
import { FieldContainer } from '../field-container/field-container';
import { inputsStyle } from '../text-input/styles/inputs';

const StyledTextArea = styled.textarea`
    ${props => inputsStyle(props.theme)}
    min-height: 6.5rem;
    min-width: 100%;
    outline: none;
    overflow: auto;
    resize: vertical;
`;

export interface TextAreaProps {
    label: string;
    defaultValue?: string;
    disabled?: boolean;
    placeholder?: string;
    required?: boolean;
    /**
    * Message displayed in case of validation error
    * @default This text area input is invalid
    **/
    validationErrorMessage?: string;
    /** Only use if you want to control input value externally */
    value?: string;

    onBlur?(event: FocusEvent<HTMLTextAreaElement>): void;
    onChange?(event: ChangeEvent<HTMLTextAreaElement>): void;
    onFocus?(event: FocusEvent<HTMLTextAreaElement>): void;
}

interface ValidityProps {
    validity: boolean;
}

export function TextArea({ onBlur, onChange, onFocus, ...props }: TextAreaProps): ReactElement {
    const [{ validity }, setValidity] = useState<ValidityProps>({ validity: true });
    const id = uuid();

    function handleBlur(event: FocusEvent<HTMLTextAreaElement>): void {
        setValidity({ validity: event.currentTarget.checkValidity() });

        if (onBlur) {
            onBlur(event);
        }
    }

    function handleChange(event: ChangeEvent<HTMLTextAreaElement>): void {
        if (onChange) {
            onChange(event);
        }
    }

    function handleFocus(event: FocusEvent<HTMLTextAreaElement>): void {
        if (onFocus) {
            onFocus(event);
        }
    }

    const { defaultValue, disabled, label, placeholder, required, validationErrorMessage, value } = props;

    return (
        <FieldContainer
            fieldId={id}
            label={label}
            valid={validity}
            validationErrorMessage={validationErrorMessage || 'This text area input is invalid'}
        >
            <StyledTextArea
                defaultValue={defaultValue}
                disabled={disabled}
                id={id}
                onBlur={handleBlur}
                onChange={handleChange}
                onFocus={handleFocus}
                placeholder={placeholder}
                required={required}
                value={value}
            />
        </FieldContainer>
    );
}
