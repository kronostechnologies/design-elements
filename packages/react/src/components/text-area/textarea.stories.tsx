import React from 'react';
import { TextArea } from './text-area';

export default {
    title: 'TextArea',
    component: TextArea,
};

export const normal = () => (
    <TextArea
        label="Text area label"
        placeholder="Enter your text here"
        validationErrorMessage="Error message"
        disabled={false}
        required={false}
    />
);
export const controlledValue = () => (
    <TextArea
        label="Text area label"
        placeholder="Enter your text here"
        value="This is the value"
        validationErrorMessage="Error message"
        disabled={false}
        required={false}
    />
);
export const eventCallbacks = () => (
    <TextArea
        label="Text area label"
        onChange={event => {
            console.log(`Custom function called on change. Current value: ${event.currentTarget.value}`);
        }}
        onBlur={event => {
            console.log(`Custom function called on blur. Current value: ${event.currentTarget.value}`);
        }}
        onFocus={event => {
            console.log(`Custom function called on focus. Current value: ${event.currentTarget.value}`);
        }}
        placeholder="Enter your text here"
        validationErrorMessage="Error message"
        disabled={false}
        required={false}
    />
);
export const required = () => (
    <TextArea
        label="Text area label"
        placeholder="Enter your text here"
        validationErrorMessage="Error message"
        disabled={false}
        required={true}
    />
);
export const defaultValue = () => (
    <TextArea
        label="Text area label"
        placeholder="Enter your text here"
        defaultValue={'Nullam eu ante vel est convallis dignissim. Fusce suscipit, wisi nec facilisis facilisis, est dui fermentum leo, quis tempor ligula erat quis odio.'}
        validationErrorMessage="Error message"
        disabled={false}
        required={false}
    />
);
export const disabled = () => (
    <TextArea
        label="A label for the disabled text area"
        placeholder="This field is disabled"
        validationErrorMessage="Error message"
        disabled={true}
        required={false}
    />
);
