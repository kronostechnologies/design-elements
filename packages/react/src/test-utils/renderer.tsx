import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
