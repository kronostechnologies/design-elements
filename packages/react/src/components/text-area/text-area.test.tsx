import { shallow } from 'enzyme';
import { doNothing } from '../../test-utils/callbacks';
import { findByTestId, getByTestId } from '../../test-utils/enzyme-selectors';
import { mountWithTheme, renderWithTheme } from '../../test-utils/renderer';
import { TextArea } from './text-area';

jest.mock('../../utils/uuid');

describe('TextArea', () => {
    const defaultProps = {
        defaultValue: 'foo',
        label: 'Comment',
        placeholder: 'Enter your comment',
        required: true,
        validationErrorMessage: 'Error message',
    };
    test('onChange callback is called when content is changed', () => {
        const callback = jest.fn();
        const wrapper = mountWithTheme(<TextArea onChange={callback} {...defaultProps} />);

        wrapper.find('textarea').simulate('change', { target: { value: 'bar' } });

        expect(callback).toHaveBeenCalledTimes(1);
    });

    test('onBlur callback is called when content is blurred', () => {
        const callback = jest.fn();
        const wrapper = mountWithTheme(<TextArea onBlur={callback} {...defaultProps} />);

        wrapper.find('textarea').simulate('blur');

        expect(callback).toHaveBeenCalledTimes(1);
    });

    test('onFocus callback is called when content is focused', () => {
        const callback = jest.fn();
        const wrapper = mountWithTheme(<TextArea onFocus={callback} {...defaultProps} />);

        wrapper.find('textarea').simulate('focus');

        expect(callback).toHaveBeenCalledTimes(1);
    });

    test('onFocus callback cannot be called when disabled', () => {
        const callback = jest.fn();
        const wrapper = mountWithTheme(<TextArea onBlur={callback} {...defaultProps} disabled />);

        wrapper.find('textarea').simulate('focus');

        expect(callback).toHaveBeenCalledTimes(0);
    });

    test('onFocus callback cannot be called when disabled', () => {
        const callback = jest.fn();
        const wrapper = mountWithTheme(<TextArea onBlur={callback} {...defaultProps} disabled />);

        wrapper.find('textarea').simulate('focus');

        expect(callback).toHaveBeenCalledTimes(0);
    });

    test('Matches the snapshot', () => {
        const tree = renderWithTheme(
            <TextArea data-testid="some-data-testid" onChange={doNothing} onBlur={doNothing} {...defaultProps} />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('Is disabled', () => {
        const tree = renderWithTheme(<TextArea onChange={doNothing} onBlur={doNothing} {...defaultProps} disabled />);

        expect(tree).toMatchSnapshot();
    });

    test('should flag input as invalid when length exceeds maxLength', () => {
        const wrapper = shallow(
            <TextArea label="Test Input" maxLength={20} />,
        );

        getByTestId(wrapper, 'textarea')
            .simulate('change', { currentTarget: { value: 'This text is longer than input length' } });

        const container = getByTestId(wrapper, 'container');
        const charCounter = getByTestId(wrapper, 'char-counter');
        expect(container.prop('valid')).toBe(false);
        expect(charCounter.prop('valid')).toBe(false);
    });

    test('should display character count when maxLength is provided', () => {
        const wrapper = shallow(
            <TextArea label="Test Input" defaultValue="Default input text" maxLength={20} />,
        );

        const charCounter = getByTestId(wrapper, 'char-counter');
        expect(charCounter.text()).toContain('18/20');
    });

    test('should not display character count when maxLength is not provided', () => {
        const wrapper = shallow(
            <TextArea label="Test Input" defaultValue="Default input text" />,
        );

        const charCounter = findByTestId(wrapper, 'char-counter');
        expect(charCounter.exists()).toBe(false);
    });
});
