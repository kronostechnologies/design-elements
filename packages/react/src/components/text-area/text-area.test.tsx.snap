// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea Is disabled 1`] = `
.c2 {
  color: #1B1C1E;
  display: block;
  font-size: 0.75rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.25rem;
  margin: 0;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

input + .c1 {
  margin-left: var(--spacing-half);
}

.c0 {
  margin: 0 0 var(--spacing-3x);
}

.c0 input,
.c0 select,
.c0 textarea {
  border-color: #60666E;
}

.c0 > :nth-child(1) {
  margin-bottom: var(--spacing-half);
}

.c3 {
  background: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  color: #000000;
  font-family: inherit;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.015rem;
  -moz-letter-spacing: 0.015rem;
  -ms-letter-spacing: 0.015rem;
  letter-spacing: 0.015rem;
  line-height: 1.5rem;
  margin: 0;
  min-height: var(--size-2x);
  outline: none;
  padding: 0 var(--spacing-1x);
  width: 100%;
  min-height: 6.5rem;
  min-width: 100%;
  outline: none;
  overflow: auto;
  resize: vertical;
}

.c3::-webkit-input-placeholder {
  color: #60666E;
}

.c3::-moz-placeholder {
  color: #60666E;
}

.c3:-ms-input-placeholder {
  color: #60666E;
}

.c3::placeholder {
  color: #60666E;
}

.c3:disabled {
  background-color: #F1F2F2;
  border-color: #DBDEE1;
  color: #DBDEE1;
}

.c3:disabled,
.c3:disabled::-webkit-input-placeholder {
  color: #B7BBC2;
}

.c3:disabled,
.c3:disabled::-moz-placeholder {
  color: #B7BBC2;
}

.c3:disabled,
.c3:disabled:-ms-input-placeholder {
  color: #B7BBC2;
}

.c3:disabled,
.c3:disabled::placeholder {
  color: #B7BBC2;
}

.c3 {
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c3:focus {
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

<div
  class="c0"
  data-testid="container"
>
  <label
    class="c1 c2"
    for="uuid1"
    id="uuid1_label"
  >
    Comment (required)
  </label>
  <textarea
    class="c3"
    data-testid="textarea"
    disabled=""
    id="uuid1"
    placeholder="Enter your comment"
    required=""
  >
    foo
  </textarea>
</div>
`;

exports[`TextArea Matches the snapshot 1`] = `
.c2 {
  color: #1B1C1E;
  display: block;
  font-size: 0.75rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.25rem;
  margin: 0;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

input + .c1 {
  margin-left: var(--spacing-half);
}

.c0 {
  margin: 0 0 var(--spacing-3x);
}

.c0 input,
.c0 select,
.c0 textarea {
  border-color: #60666E;
}

.c0 > :nth-child(1) {
  margin-bottom: var(--spacing-half);
}

.c3 {
  background: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  color: #000000;
  font-family: inherit;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.015rem;
  -moz-letter-spacing: 0.015rem;
  -ms-letter-spacing: 0.015rem;
  letter-spacing: 0.015rem;
  line-height: 1.5rem;
  margin: 0;
  min-height: var(--size-2x);
  outline: none;
  padding: 0 var(--spacing-1x);
  width: 100%;
  min-height: 6.5rem;
  min-width: 100%;
  outline: none;
  overflow: auto;
  resize: vertical;
}

.c3::-webkit-input-placeholder {
  color: #60666E;
}

.c3::-moz-placeholder {
  color: #60666E;
}

.c3:-ms-input-placeholder {
  color: #60666E;
}

.c3::placeholder {
  color: #60666E;
}

.c3:disabled {
  background-color: #F1F2F2;
  border-color: #DBDEE1;
  color: #DBDEE1;
}

.c3:disabled,
.c3:disabled::-webkit-input-placeholder {
  color: #B7BBC2;
}

.c3:disabled,
.c3:disabled::-moz-placeholder {
  color: #B7BBC2;
}

.c3:disabled,
.c3:disabled:-ms-input-placeholder {
  color: #B7BBC2;
}

.c3:disabled,
.c3:disabled::placeholder {
  color: #B7BBC2;
}

.c3 {
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c3:focus {
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

<div
  class="c0"
  data-testid="container"
>
  <label
    class="c1 c2"
    for="uuid1"
    id="uuid1_label"
  >
    Comment (required)
  </label>
  <textarea
    class="c3"
    data-testid="some-data-testid"
    id="uuid1"
    placeholder="Enter your comment"
    required=""
  >
    foo
  </textarea>
</div>
`;
